### PR TITLE
Lua settings improvements

### PR DESCRIPTION
**Pull request by PilzAdam**
Changes:
- Key settings are added to `minetest.conf.example`
- Added missing key and noclip settings to `settingtypes.txt
- Expanded description on key settings
- Fixed parallax typos
- `flags` now fully supported with a list of possible values
- Added input validation for `int`, `float` and `flags` types (type safety and min / max checks)
- Restructured graphics section
- Additional `settingtypes.txt` files are passed from games and mods root folders; they are displayed in special categories.
- Added description for format of `noise_params` type

### CODE REVIEW

**Comment by kilbith** [comment]
@PilzAdam L.267, please change "View" to "Viewing" like the setting above.

**Comment by kilbith** [comment]
@PilzAdam You've cancelled (at least) 9d21fa3 with 5675486.

**Comment by PilzAdam** [comment]
@kilbith Thanks for spotting that; I check everything else now.

**Comment by PilzAdam** [comment]
Everything important is now done here, except the support for grouped noise param settings, but this is due to a missing Lua API functionality for grouped settings (or structs).

The only things missing from the list in #3266 are fancier input dialogs for certain types, but that's not so high priority.

I think this can be merged now.

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -29,6 +31,11 @@
 #            - default (if default is not specified then "" is set)
 #   * key:
 #            - default
+#   * flags:
+#            Flags are always comman seperated without spaces.
```
Comment (est31) [inline builtin/settingtypes.txt]: comman?

File: builtin/settingtypes.txt [hunk 2]
```diff
@@ -430,40 +583,41 @@ show_debug (Show debug info) bool false
...
 
 #    Network port to listen (UDP).
+#    This value will be overridden when starting from the main menu.
 port (Server port) int 30000
 
 #    The network interface that the server listens on.
 bind_address (Bind address) string
 
-#    Set to true to disallow old clients from connecting.
+#    Set to true to disallow old, possibly not fully compatible clients from connecting.
```
Comment (est31) [inline builtin/settingtypes.txt]: What does this have to do with "full compatibility"? If we support a network protocol version, we _are_ fully compatible to it.

File: builtin/settingtypes.txt [hunk 3]
```diff
@@ -619,20 +770,21 @@ profiler_print_interval (Profiling print interval) int 0
...
-#    http://www.sqlite.org/pragma.html#pragma_synchronous only numeric values: 0 1 2
-sqlite_synchronous (Synchronous SQLite) int 2 0 2
+#    See http://www.sqlite.org/pragma.html#pragma_synchronous
+sqlite_synchronous (Synchronous SQLite) enum 2 0,1,2
 
 #    Length of a server tick and the interval at which objects are generally updated over network.
 dedicated_server_step (Dedicated server step) float 0.1
 
-#    Can be set to true to disable shutting down on invalid world data.
+#    Can be enabled to disable shutting down on invalid world data.
```
Comment (est31) [inline builtin/settingtypes.txt]: Perhaps it would be easier to understand if phrased "If enabled, invalid world data won't cause the server to shut down"?

File: builtin/settingtypes.txt [hunk 4]
```diff
@@ -789,36 +938,38 @@ name (Player name) string
...
-#         - verbose
-debug_log_level (Debug log level) enum action warning,action,info,verbose
+#    -    <nothing> (no logging)
+#    -    none (messages with no level)
+#    -    error
+#    -    warning
+#    -    action
+#    -    info
+#    -    verbose
+debug_log_level (Debug log level) enum action ,warning,action,info,verbose
```
Comment (est31) [inline builtin/settingtypes.txt]: why the leading comma?

File: builtin/mainmenu/tab_settings.lua [hunk 5]
```diff
@@ -190,22 +234,77 @@ local function parse_config_file()
...
+		local games_category_initialized = false
+		local index = 1
+		local game = gamemgr.get_game(index)
+		while game do
+			local file = io.open(game.path .. DIR_DELIM .. FILENAME, "r")
+			if file then
+				if not games_category_initialized then
+					table.insert(settings, {
+						name = "Games",
+						translation = fgettext_ne("Games"), -- not used, but needed for xgettext
```
Comment (est31) [inline builtin/mainmenu/tab_settings.lua]: Perhaps make it more obvious that it really isn't needed, eg by making `translation` a local variable?

**Comment by est31** [comment]
Really great idea to scan for `settingstypes.txt` in the mods and game directories as well.

**Comment by est31** [comment]
For modders and game creators: should we add documentation for how to automatically generate minetest.conf.example for mods/games from settingstypes.txt?

File: builtin/settingtypes.txt (hunk 4 above)
Comment (PilzAdam) [inline builtin/settingtypes.txt]: This adds an empty entry to the enum

> `<nothing> (no logging)`

File: builtin/settingtypes.txt (hunk 2 above)
Comment (PilzAdam) [inline builtin/settingtypes.txt]: Maybe not from the server admin's perspective. Older clients may be fully compatible from a network point of view, but they may not support all the new features that the server expects.

File: builtin/settingtypes.txt (hunk 1 above)
Comment (PilzAdam) [inline builtin/settingtypes.txt]: Typo; will fix.

File: builtin/settingtypes.txt (hunk 3 above)
Comment (PilzAdam) [inline builtin/settingtypes.txt]: Will fix.

File: builtin/mainmenu/tab_settings.lua (hunk 5 above)
Comment (PilzAdam) [inline builtin/mainmenu/tab_settings.lua]: Will fix.

**Comment by PilzAdam** [comment]
@est31 addressed all your remarks.

If we want to expose the generation of `minetest.conf.example` then it should be done with a command line option or an extra button in the GUI. Editing `builtin/mainmenu/tab_settings.lua` is not an option for end-users (on some distros they may not be able to edit this file, since it's installed in the system-wide part).

I would personally prefer a way to pass command-line options to the main menu (maybe with registering them first, so they can be printed in `--help`), but that may be too complicated for this use-case.

File: builtin/settingtypes.txt (hunk 4 above)
Comment (est31) [inline builtin/settingtypes.txt]: I see.

File: builtin/settingtypes.txt [hunk 6]
```diff
@@ -49,12 +58,18 @@
...
-#    Fast movement (keymap_special1).
+#    Fast movement (via use key).
+#    This requires the "fast" privilege on the server.
 fast_move (Fast movement) bool false
 
+#    If enabled together with fly mode, player is able to fly through solid nodes.
+#    This requires the "noclip" privilege on the server.
+noclip (Noclip) bool false
+
 #    Smooths camera when moving and looking arround.
```
Comment (rubenwardy) [inline builtin/settingtypes.txt]: "arround"

File: builtin/settingtypes.txt [hunk 7]
```diff
@@ -281,63 +308,181 @@ texture_clean_transparent (Clean transparent textures) bool false
...
+#    Set this to be equal to viewing range minimum to disable the auto-adjustment algorithm.
+viewing_range_nodes_min (Viewing range minimum) int 35
+
+#    Vertical initial window size.
+screenW (Screen width) int 800
+
+#    Horizontal initial window size.
+screenH (Screen height) int 600
+
+#    Fullscreen mode.
```
Comment (rubenwardy) [inline builtin/settingtypes.txt]: I'd argue fullscreen isn't advanced.

File: builtin/settingtypes.txt [hunk 8]
```diff
@@ -490,38 +648,35 @@ max_packets_per_iteration (Max. packets per iteration) int 1024
...
 default_game (Default game) string minetest
 
-#    Message of the Day.
-motd (Message of the day) string Welcome to this awesome Minetest server!
+#    Message of the day displayed to players connecting.
+motd (Message of the day) string
 
-#    Maximum number of players connected simultaneously.
+#    Maximum number of players that can connect simultaneously.
 max_users (Maximum users) int 15
```
Comment (rubenwardy) [inline builtin/settingtypes.txt]: I'd argue this is a server setting, rather than a game setting: same for motd.

"game" for me means settings related to gameplay.

File: builtin/settingtypes.txt [hunk 9]
```diff
@@ -660,22 +816,21 @@ max_block_generate_distance (Max block generate distance) int 6
...
+#    -    The map generator works in groups of 80x80x80 nodes (5x5x5 MapBlocks).
+#    -    Those groups have an offset of -32, -32 nodes from the origin.
+#    -    Only groups which are within the map_generation_limit are generated
 map_generation_limit (Map generation limit) int 31000 0 31000
 
 #    Global map generation attributes.
-#    Currently supported: trees, caves, flat, dungeons, light.
 #    Flags that are not specified in the flag string are not modified from the default.
-#    To explicitly turn off a flag, prepend "no" to the beginning, e.g. nolight.
+#    Flags starting with "no" are used to explicitly disable them.
```
Comment (rubenwardy) [inline builtin/settingtypes.txt]: The example was useful.

File: builtin/settingtypes.txt [hunk 10]
```diff
@@ -780,45 +931,49 @@ secure.enable_security (Enable mod security) bool false
...
 #    When running a server, clients connecting with this name are admins.
+#    When starting from the main menu, this is overridden.
 name (Player name) string
 
-#    Override language. When no value is provided (default) system language is used.
-#    Check "locale" directory for the list of available translations.
-language (Language) string
+#    Set the language. Leave empty to use the system language.
+#    A restart is required after changing this.
+language (Language) enum   ,be,cs,da,de,eo,es,et,fr,hu,id,it,ja,jbo,ko,ky,lt,nb,nl,pl,pt,pt_BR,ro,ru,tr,uk,zh_CN,zh_TW
```
Comment (rubenwardy) [inline builtin/settingtypes.txt]: What about en?

File: src/settings_translation_file.cpp [hunk 11]
```diff
@@ -22,42 +24,168 @@ fake_function() {
...
+	gettext("HUD toggle key");
+	gettext("Key for toggling the display of the HUD.\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
+	gettext("Chat toggle key");
+	gettext("Key for toggling the display of the chat.\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
+	gettext("Fog toggle key");
+	gettext("Key for toggling the display of the fog.\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
+	gettext("Camera update toggle key");
+	gettext("Key for toggling the camrea update. Only used for development\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
+	gettext("Debug info toggle key");
+	gettext("Key for toggling the display of debug info.\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
```
Comment (rubenwardy) [inline src/settings_translation_file.cpp]: as you can't click links, you should make a tinyurl style shortcut, or a wiki link.

**Comment by paramat** [comment]
New notes for map_generation_limit for settingtypes.txt and minetest.conf.example.
The unnecessary 'please note' line can be removed and the notes formatted as elsewhere in these files.

Map generation is limited to this distance in nodes from (0, 0, 0), along all axes in positive and negative directions. 
Value is limited to 31000.
By default mapgen generates in chunks of 80x80x80 nodes.
By default these chunks are offset (-32, -32, -32) nodes from the origin.
Only chunks within the map_generation_limit are generated.

File: src/settings_translation_file.cpp (hunk 11 above)
Comment (PilzAdam) [inline src/settings_translation_file.cpp]: These settings are not visible in GUI, only in the generated mintest.conf.example.

**Comment by PilzAdam** [comment]
Thanks for all your suggestions on `settingtypes.txt`, but I'd like to first merge this PR, since it contains important functionality. Improving `settingtypes.txt` will probably take a long time and should be subject of future PRs.

**Comment by paramat** [comment]
Cool. For the moment don't worry too much about the setting of noise parameters, as this is advanced stuff that can always be done using minetest.conf.example.

**Comment by est31** [comment]
Don't edit that file anymore. If you want to change something in minetest.conf.example, edit settingstypes.txt instead, and @PilzAdam will generate minetest.conf.example with automated tools.
